File: plugins/plugin-knowledge/python/elizaos_plugin_knowledge/actions.py
import re
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Optional, List

KNOWLEDGE_KEYWORDS = [
    "process",
    "add",
    "upload",
    "document",
    "knowledge",
    "learn",
    "remember",
    "store",
    "ingest",
    "file",
]

SEARCH_KEYWORDS = ["search", "find", "look up", "query", "what do you know about"]

KNOWLEDGE_SEARCH_KEYWORDS = ["knowledge", "information", "document", "database"]


@dataclass
class ActionContext:
    message: dict = field(default_factory=dict)
    agent_id: str = ""
    room_id: Optional[str] = None
    entity_id: Optional[str] = None
    state: dict = field(default_factory=dict)


class KnowledgeAction(ABC):
    @property
    @abstractmethod
    def name(self) -> str:
        pass

    @property
    @abstractmethod
    def description(self) -> str:
        pass

    @abstractmethod
    def validate(self, context: ActionContext) -> bool:
        pass

    @abstractmethod
    async def execute(self, context: ActionContext) -> dict:
        pass


class ProcessKnowledgeAction(KnowledgeAction):
    @property
    def name(self) -> str:
        return "PROCESS_KNOWLEDGE"

    @property
    def description(self) -> str:
        return (
            "Process and store knowledge from a file path or text content into the knowledge base"
        )

    def validate(self, context: ActionContext) -> bool:
        text = (
            context.message.get("content", {}).get("text", "") or context.message.get("text", "")
        ).lower()

        has_keyword = any(keyword in text for keyword in KNOWLEDGE_KEYWORDS)
        has_path = "/" in text and "http" not in text

        return has_keyword or has_path

    async def execute(self, context: ActionContext) -> dict:
        text = context.message.get("content", {}).get("text", "") or context.message.get("text", "")

        path_pattern = r"(?:/[\w.\-]+)+|(?:[a-zA-Z]:[/\\][\w\s.\-]+(?:[/\\][\w\s.\-]+)*)"
        path_match = re.search(path_pattern, text)

        if path_match:
            file_path = path_match.group()
            return {
                "action": self.name,
                "mode": "file",
                "file_path": file_path,
                "agent_id": context.agent_id,
                "room_id": context.room_id,
                "entity_id": context.entity_id,
                "status": "pending",
                "message": f"Processing document at {file_path}",
            }
        else:
            knowledge_content = text.strip()

            if not knowledge_content:
                return {
                    "action": self.name,
                    "success": False,
                    "error": "No content provided to process",
                }

            return {
                "action": self.name,
                "mode": "text",
                "content": knowledge_content,
                "agent_id": context.agent_id,
                "room_id": context.room_id,
                "entity_id": context.entity_id,
                "status": "pending",
                "message": "Processing text content",
            }


class SearchKnowledgeAction(KnowledgeAction):
    @property
    def name(self) -> str:
        return "SEARCH_KNOWLEDGE"

    @property
    def description(self) -> str:
        return "Search the knowledge base for specific information"

    def validate(self, context: ActionContext) -> bool:
        text = (
            context.message.get("content", {}).get("text", "") or context.message.get("text", "")
        ).lower()

        has_search_keyword = any(keyword in text for keyword in SEARCH_KEYWORDS)
        has_knowledge_keyword = any(keyword in text for keyword in KNOWLEDGE_SEARCH_KEYWORDS)

        return has_search_keyword and has_knowledge_keyword

    async def execute(self, context: ActionContext) -> dict:
        text = context.message.get("content", {}).get("text", "") or context.message.get("text", "")

        query = text.lower()
        for word in [
            "search",
            "find",
            "look up",
            "query",
            "your",
            "my",
            "knowledge",
            "base",
            "for",
            "information",
            "document",
            "database",
        ]:
            query = query.replace(word, "")
        query = query.strip()

        if not query:
            return {
                "action": self.name,
                "success": False,
                "error": "No search query provided",
                "message": "What would you like me to search for in my knowledge base?",
            }

        return {
            "action": self.name,
            "query": query,
            "agent_id": context.agent_id,
            "room_id": context.room_id,
            "entity_id": context.entity_id,
            "status": "pending",
            "message": f"Searching knowledge base for: {query}",
        }


def get_actions() -> List[KnowledgeAction]:
    return [ProcessKnowledgeAction(), SearchKnowledgeAction()]


process_knowledge_action = ProcessKnowledgeAction()
search_knowledge_action = SearchKnowledgeAction()
knowledge_actions = [process_knowledge_action, search_knowledge_action]

File: packages/typescript/src/features/knowledge/actions.ts
import * as fs from "node:fs";
import * as path from "node:path";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../i18n/validation-keywords.ts";
import { logger } from "../../logger";
import type {
	Action,
	Content,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
	UUID,
} from "../../types";
import { addKnowledgeFromFilePath } from "./docs-loader.ts";
import { KnowledgeService } from "./service.ts";
import { createKnowledgeNoteFilename, deriveKnowledgeTitle } from "./utils.ts";

type ExtendedValidator = (
	runtime: IAgentRuntime,
	message: Memory,
	state?: State,
	options?: unknown,
) => Promise<boolean>;

const PROCESS_KNOWLEDGE_TERMS = getValidationKeywordTerms(
	"action.processKnowledge.request",
	{
		includeAllLocales: true,
	},
);
const SEARCH_KNOWLEDGE_TERMS = getValidationKeywordTerms(
	"action.searchKnowledge.request",
	{
		includeAllLocales: true,
	},
);
const KNOWLEDGE_PATH_PATTERN =
	/(?:\/[\w.-]+)+|(?:[a-zA-Z]:[\\/][\w\s.-]+(?:[\\/][\w\s.-]+)*)/;

export const processKnowledgeAction: Action = {
	name: "PROCESS_KNOWLEDGE",
	description:
		"Process and store knowledge from a file path or text content into the knowledge base",

	similes: [],

	examples: [
		[
			{
				name: "user",
				content: {
					text: "Process the document at /path/to/document.pdf",
				},
			},
			{
				name: "assistant",
				content: {
					text: "I'll process the document at /path/to/document.pdf and add it to my knowledge base.",
					actions: ["PROCESS_KNOWLEDGE"],
				},
			},
		],
		[
			{
				name: "user",
				content: {
					text: "Add this to your knowledge: The capital of France is Paris.",
				},
			},
			{
				name: "assistant",
				content: {
					text: "I'll add that information to my knowledge base.",
					actions: ["PROCESS_KNOWLEDGE"],
				},
			},
		],
	],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: unknown,
	): Promise<boolean> => {
		const __avLegacyValidate: ExtendedValidator = async (
			runtime: IAgentRuntime,
			message: Memory,
			_state?: State,
			_options?: unknown,
		) => {
			const text = message.content.text ?? "";
			const hasKeyword =
				findKeywordTermMatch(text, PROCESS_KNOWLEDGE_TERMS) !== undefined;
			const hasPath = KNOWLEDGE_PATH_PATTERN.test(text);
			const service = runtime.getService(KnowledgeService.serviceType);
			if (!service) {
				logger.warn(
					"Knowledge service not available for PROCESS_KNOWLEDGE action",
				);
				return false;
			}

			return hasKeyword || hasPath;
		};
		try {
			return Boolean(
				await __avLegacyValidate(runtime, message, state, options),
			);
		} catch {
			return false;
		}
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	) => {
		try {
			const service = runtime.getService<KnowledgeService>(
				KnowledgeService.serviceType,
			);
			if (!service) {
				throw new Error("Knowledge service not available");
			}

			const text = message.content.text || "";
			const pathMatch = text.match(KNOWLEDGE_PATH_PATTERN);

			let response: Content;

			if (pathMatch) {
				const filePath = pathMatch[0];

				if (!fs.existsSync(filePath)) {
					response = {
						text: `I couldn't find the file at ${filePath}. Please check the path and try again.`,
					};

					if (callback) {
						await callback(response);
					}
					return;
				}

				const fileName = path.basename(filePath);
				const result = await addKnowledgeFromFilePath({
					service,
					agentId: runtime.agentId,
					worldId: runtime.agentId,
					roomId: runtime.agentId,
					entityId: runtime.agentId,
					filePath,
					metadata: {
						source: "learned",
						learnedVia: "PROCESS_KNOWLEDGE",
						learnedFromPath: filePath,
					},
				});

				response = {
					text: `I've successfully processed the document "${fileName}". It has been split into ${result?.fragmentCount || 0} searchable fragments and added to my knowledge base.`,
				};
			} else {
				const knowledgeContent = text
					.replace(
						/^(add|store|remember|process|learn)\s+(this|that|the following)?:?\s*/i,
						"",
					)
					.trim();

				if (!knowledgeContent) {
					response = {
						text: "I need some content to add to my knowledge base. Please provide text or a file path.",
					};

					if (callback) {
						await callback(response);
					}
					return;
				}

				const title = deriveKnowledgeTitle(
					knowledgeContent,
					"Learned knowledge",
				);
				const filename = createKnowledgeNoteFilename(title);
				const knowledgeOptions = {
					clientDocumentId: "" as UUID,
					contentType: "text/plain",
					originalFilename: filename,
					worldId: runtime.agentId,
					content: knowledgeContent,
					roomId: runtime.agentId,
					entityId: runtime.agentId,
					metadata: {
						source: "learned",
						learnedVia: "PROCESS_KNOWLEDGE",
						title,
						filename,
						originalFilename: filename,
						fileExt: "txt",
						fileType: "text/plain",
						contentType: "text/plain",
						fileSize: Buffer.byteLength(knowledgeContent, "utf8"),
						textBacked: true,
					},
				};

				await service.addKnowledge(knowledgeOptions);

				response = {
					text: `I've added that information to my knowledge base. It has been stored and indexed for future reference.`,
				};
			}

			if (callback) {
				await callback(response);
			}
			return { success: true, text: response.text };
		} catch (error) {
			logger.error({ error }, "Error in PROCESS_KNOWLEDGE action");

			const errorResponse: Content = {
				text: `I encountered an error while processing the knowledge: ${error instanceof Error ? error.message : String(error)}`,
			};

			if (callback) {
				await callback(errorResponse);
			}
			return {
				success: false,
				error: error instanceof Error ? error.message : String(error),
			};
		}
	},
};

export const searchKnowledgeAction: Action = {
	name: "SEARCH_KNOWLEDGE",
	description: "Search the knowledge base for specific information",

	similes: [
		"search knowledge",
		"find information",
		"look up",
		"query knowledge base",
		"search documents",
		"find in knowledge",
	],

	examples: [
		[
			{
				name: "user",
				content: {
					text: "Search your knowledge for information about quantum computing",
				},
			},
			{
				name: "assistant",
				content: {
					text: "I'll search my knowledge base for information about quantum computing.",
					actions: ["SEARCH_KNOWLEDGE"],
				},
			},
		],
	],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: unknown,
	): Promise<boolean> => {
		const __avLegacyValidate: ExtendedValidator = async (
			runtime: IAgentRuntime,
			message: Memory,
			_state?: State,
			_options?: unknown,
		) => {
			const text = message.content.text ?? "";
			const hasSearchKeyword =
				findKeywordTermMatch(text, SEARCH_KNOWLEDGE_TERMS) !== undefined;
			const service = runtime.getService(KnowledgeService.serviceType);
			if (!service) {
				return false;
			}

			return hasSearchKeyword;
		};
		try {
			return Boolean(
				await __avLegacyValidate(runtime, message, state, options),
			);
		} catch {
			return false;
		}
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	) => {
		try {
			const service = runtime.getService<KnowledgeService>(
				KnowledgeService.serviceType,
			);
			if (!service) {
				throw new Error("Knowledge service not available");
			}

			const text = message.content.text || "";

			const query = text
				.replace(
					/^(search|find|look up|query)\s+(your\s+)?knowledge\s+(base\s+)?(for\s+)?/i,
					"",
				)
				.trim();

			if (!query) {
				const response: Content = {
					text: "What would you like me to search for in my knowledge base?",
				};

				if (callback) {
					await callback(response);
				}
				return;
			}

			const searchMessage: Memory = {
				...message,
				content: {
					text: query,
				},
			};

			const results = await service.getKnowledge(searchMessage);

			let response: Content;

			if (results.length === 0) {
				response = {
					text: `I couldn't find any information about "${query}" in my knowledge base.`,
				};
			} else {
				const formattedResults = results
					.slice(0, 3)
					.map((item, index) => `${index + 1}. ${item.content.text}`)
					.join("\n\n");

				response = {
					text: `Here's what I found about "${query}":\n\n${formattedResults}`,
				};
			}

			if (callback) {
				await callback(response);
			}
			return { success: true, text: response.text };
		} catch (error) {
			logger.error({ error }, "Error in SEARCH_KNOWLEDGE action");

			const errorResponse: Content = {
				text: `I encountered an error while searching the knowledge base: ${error instanceof Error ? error.message : String(error)}`,
			};

			if (callback) {
				await callback(errorResponse);
			}
			return {
				success: false,
				error: error instanceof Error ? error.message : String(error),
			};
		}
	},
};

export const knowledgeActions = [processKnowledgeAction, searchKnowledgeAction];
